src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp

```diff
@@ -0,0 +1,730 @@
+            p_end);
+
+        pb_desc->AddParam(
+            param_id++, L"bump_up_vector", TYPE_INT, 0, IDS_BUMP_UP_VECTOR,
+            p_ui, TYPE_INT_COMBOBOX, IDC_COMBO_BUMP_UP_VECTOR,
+            2, IDS_COMBO_BUMP_UP_VECTOR_Y, IDS_COMBO_BUMP_UP_VECTOR_Z,
+            p_vals, 0, 1,
+            p_default, 1,
+            p_end);
+    }
+
+    MaxParam add_output_parameter(
+        ParamBlockDesc2*                    pb_desc,
+        const std::vector<OSLParamInfo>&    output_params,
+        IdNameMap&                          string_map,
+        int&                                param_id,
+        int&                                ctrl_id)
+    {
+        int string_id = ctrl_id++;
+        string_map.insert(std::make_pair(string_id, L"Output"));
+
+        MaxParam max_output_param;
+        max_output_param.max_label_str = "Output";
+        max_output_param.osl_param_name = "output";
+        max_output_param.param_type = MaxParam::StringPopup;
+        max_output_param.max_ctrl_id = string_id;
+        max_output_param.max_param_id = param_id;
+        max_output_param.connectable = false;
+
+        int ctrl_id_1 = ctrl_id++;
+        int p_id = param_id++;
+
+        pb_desc->AddParam(
+            p_id,
+            L"Output",
+            TYPE_INT,
+            0,
+            string_id,
+            p_ui, TYPE_INT_COMBOBOX, ctrl_id_1,
+            0,
+            p_end
+        );
+
+        int num_items = static_cast<int>(output_params.size());
+
+        Tab<int> string_items;
+        string_items.SetCount(num_items);
+        for (size_t i = 0, e = num_items; i < e; ++i)
+        {
+            int str_id = ctrl_id++;
+            string_items[i] = str_id;
+
+            string_map.insert(std::make_pair(str_id, utf8_to_wide(output_params[i].paramName)));
+        }
+
+        pb_desc->ParamOptionContentValues(p_id, string_items);
+
+        return max_output_param;
+    }
+}
+
+void OSLShaderRegistry::create_class_descriptors()
+{
+    registerShadingNodes(m_shaders);
+
+    for (auto& shader : m_shaders)
+    {
+        ClassDesc2* class_descr(new OSLPluginClassDesc(&shader));
+        ParamBlockDesc2* param_block_descr(new ParamBlockDesc2(
+            // --- Required arguments ---
+            0,                                          // parameter block's ID
+            L"oslTextureMapParams",                     // internal parameter block's name
+            0,                                          // ID of the localized name string
+            class_descr,                                // class descriptor
+            P_AUTO_CONSTRUCT,                           // block flags
+
+                                                        // --- P_AUTO_CONSTRUCT arguments ---
+            0,                                          // parameter block's reference number
+            p_end
+        ));
+
+        int ctrl_id = 100;
+        int param_id = 0;
+        for (auto& param_info : shader.m_params)
```

Comment:
Can this be made const?

Reply:
I think it can't since param_info.m_max_param get assigned values in add_parameter()